creme/linear_model/ffm.py

```diff
@@ -0,0 +1,364 @@
+            self.weight_initializer = weight_initializer
+        else:
+            self.weight_initializer = optim.initializers.Zeros()
+
+        if latent_initializer:
+            self.latent_initializer = latent_initializer
+        else:
+            self.latent_initializer = optim.initializers.Normal(sigma=.1, random_state=random_state)
+
+        self.l1 = l1
+        self.l2 = l2
+        self.clip_gradient = clip_gradient
+
+        random_latents = functools.partial(
+            self.latent_initializer,
+            shape=self.n_factors
+        )
+
+        field_latents_dict = functools.partial(
+            collections.defaultdict, random_latents
+        )
+
+        self.weights = collections.defaultdict(self.weight_initializer)
+        self.latents = collections.defaultdict(field_latents_dict)
+
+    def _field(self, j):
+        return j.split('_')[0]
+
+    def _raw_dot(self, x):
+
+        # For notational convenience
+        w0, w, v = self.intercept, self.weights, self.latents
+        field = self._field
+
+        # Start with the intercept
+        y_pred = w0
+
+        # Add the unary interactions
+        y_pred += utils.math.dot(x, w)
+
+        # Add the pairwise interactions
+        y_pred += sum(
+            x[j1] * x[j2] * np.dot(v[j1][field(j2)], v[j2][field(j1)])
+            for j1, j2 in itertools.combinations(x.keys(), 2)
+        )
+
+        return y_pred
+
+    def _latent_gradient(self, j, field, f, x):
+
+        # For notational convenience
+        _field = self._field
+
+        # Get derivative terms
+        derivative_terms = (
+            x[j] * xj2 * self.latents[j2][_field(j)][f]
+            for j2, xj2 in x.items()
+            if j2 != j and _field(j2) == field
+        )
+
+        return sum(derivative_terms)
+
+    def fit_one(self, x, y, sample_weight=1.):
+
+        # For notational convenience
+        k, l1, l2 = self.n_factors, self.l1, self.l2
+        w0, w, v = self.intercept, self.weights, self.latents
+        w0_lr = self.intercept_lr.get(self.optimizer.n_iterations)
+
+        # Some optimizers need to do something before a prediction is made
+        self.weights = self.optimizer.update_before_pred(w=w)
+
+        for j in x.keys():
+            self.latents[j] = self.optimizer.update_before_pred(w=v[j])
+
+        # Obtain the gradient of the loss with respect to the raw output
+        g_loss = self.loss.gradient(y_true=y, y_pred=self._raw_dot(x))
+
+        # Clamp the gradient to avoid numerical instability
+        g_loss = utils.math.clamp(g_loss, minimum=-self.clip_gradient, maximum=self.clip_gradient)
+
+        # Apply the sample weight
+        g_loss *= sample_weight
+
+        # Update the intercept
+        sign = lambda x: -1 if x < 0 else (1 if x > 0 else 0)
+
+        self.intercept -= w0_lr * (g_loss + l1 * sign(w0) + 2. * l2 * w0)
+
+        # Update the weights
+        gradient = {
+            j: g_loss * xj + l1 * sign(w[j]) + 2. * l2 * w[j]
```

Comment:
Do you have a reference for the L1 regularization method you're using? As we discussed it would preferable if we extracted the L1/L2 logic into separate classes. But I haven't thought it through, it might not be possible!

Reply:
As discussed earlier today, I'm just using the basic and naive method as a first step. Unfortunately it's not efficient in an online setting and results with a lot of non 0 features.

I would point to you your reference here: [L1 regularization for linear models #76](https://github.com/creme-ml/creme/issues/76).The first paper pointed explains well our problem and provide clues for improvement.

You are right about the regularization logic, that's something we need to think about. Let's do it as a next step